src/Microsoft.ApplicationInsights.AspNetCore/DiagnosticListeners/Implementation/HostingDiagnosticListener.cs

```diff
@@ -271,18 +303,27 @@ public void OnHttpRequestInStart(HttpContext httpContext)
                     currentActivity = newActivity;
                 }
 
-                var requestTelemetry = this.InitializeRequestTelemetry(httpContext, currentActivity, Stopwatch.GetTimestamp());
-                if (this.enableW3CHeaders && sourceAppId != null)
-                {
-                    requestTelemetry.Source = sourceAppId;
-                }
-
+                var requestTelemetry = this.InitializeRequestTelemetry(httpContext, currentActivity, Stopwatch.GetTimestamp(), legacyRootId);
                 requestTelemetry.Context.Operation.ParentId = originalParentId;
 
                 this.AddAppIdToResponseIfRequired(httpContext, requestTelemetry);
             }
         }
 
+        private string ExtractOperationIdFromRequestId(string originalParentId)
+        {
+            int indexPipe = originalParentId.IndexOf('|');
```

Comment:
Combine this method with `IsCompatibleW3CTraceID and name it like `TryGetW3CCompatibleId(out var result)`.
- check if first symbol is `|` and fail fast if it's not. assume start = 1 if it is
- check if 33rd symbol is `.` fail fast if it's not, assume end = 33
- check symbols between 1 and 33 for being in a-z0-9 range
- return out string and success/failure